imap/sievedir.c

```diff
@@ -0,0 +1,426 @@
+    const char *ptr;
+
+    /* must be at least one character long */
+    if (len < 1) return 0;
+
+    ptr = buf_base(name);
+
+    for (lup = 0; lup < len; lup++) {
+        if ((ptr[lup] == '/') || (ptr[lup] == '\0')) return 0;
+    }
+
+    return (lup < 1013);
+}
+
+EXPORTED int sievedir_script_exists(const char *sievedir, const char *name)
+{
+    char path[PATH_MAX];
+    struct stat sbuf;
+
+    snprintf(path, sizeof(path), "%s/%s%s", sievedir, name, SCRIPT_SUFFIX);
+
+    return (stat(path, &sbuf) == 0);
+}
+
+EXPORTED const char *sievedir_get_active(const char *sievedir)
+{
+    static char target[PATH_MAX];
+    char link[PATH_MAX];
+    ssize_t tgt_len;
+
+    snprintf(link, sizeof(link), "%s/%s", sievedir, DEFAULTBC_NAME);
+
+    tgt_len = readlink(link, target, sizeof(target) - 1);
+
+    if (tgt_len > BYTECODE_SUFFIX_LEN) {
+        target[tgt_len - BYTECODE_SUFFIX_LEN] = '\0';
+        return target;
+    }
+    else if (tgt_len == -1 && errno != ENOENT) {
+        syslog(LOG_ERR, "IOERROR: readlink(%s): %m", link);
+    }
+
+    return NULL;
+}
+
+EXPORTED int sievedir_script_isactive(const char *sievedir, const char *name)
+{
+    if (!name) return 0;
+
+    return (strcmpnull(name, sievedir_get_active(sievedir)) == 0);
+}
+
+EXPORTED int sievedir_activate_script(const char *sievedir, const char *name)
+{
+    char target[PATH_MAX];
+    char active[PATH_MAX];
+    char tmp[PATH_MAX+4];  /* +4 for ".NEW" */
+
+    if (sievedir_script_isactive(sievedir, name)) {
+        /* already active - nothing to do here */
+        return SIEVEDIR_OK;
+    }
+
+    snprintf(target, sizeof(target), "%s%s", name, BYTECODE_SUFFIX);
+    snprintf(active, sizeof(active), "%s/%s", sievedir, DEFAULTBC_NAME);
+    snprintf(tmp, sizeof(tmp), "%s.NEW", active);
+
+    /* N.B symlink() does NOT verify target for anything but string validity,
+     * so activation of a nonexistent script will report success.
+     */
+    if (symlink(target, tmp) < 0) {
+        syslog(LOG_ERR, "IOERROR: unable to symlink %s as %s: %m", target, tmp);
+        return SIEVEDIR_IOERROR;
+    }
+
+    if (rename(tmp, active) < 0) {
+        syslog(LOG_ERR, "IOERROR: unable to rename %s to %s: %m", tmp, active);
+        unlink(tmp);
+        return SIEVEDIR_IOERROR;
+    }
+
+    return SIEVEDIR_OK;
+}
+
+EXPORTED int sievedir_deactivate_script(const char *sievedir)
+{
+    char active[PATH_MAX];
+
+    snprintf(active, sizeof(active), "%s/defaultbc", sievedir);
+    if (unlink(active) != 0 && errno != ENOENT) {
+        syslog(LOG_ERR, "IOERROR: unable to unlink %s: %m", active);
+        return SIEVEDIR_IOERROR;
+    }
+
+    return SIEVEDIR_OK;
+}
+
+EXPORTED int sievedir_delete_script(const char *sievedir, const char *name)
+{
+    char path[PATH_MAX];
+    int r;
+
+    snprintf(path, sizeof(path), "%s/%s%s", sievedir, name, SCRIPT_SUFFIX);
+    r = unlink(path);
+    if (r) {
+        if (errno == ENOENT) return SIEVEDIR_NOTFOUND;
```

Comment:
So, if there is a .bc file but the corresponding .script file is missing for some reason, this function will bail out here and not remove the .bc file?

Reply:
https://github.com/cyrusimap/cyrus-imapd/commit/4a3e19ae40feab8cbfe8a27f0685e23439a27280